python/pyspark/sql/types.py

```diff
@@ -997,12 +997,13 @@ def _restore_object(dataType, obj):
     # same object in most cases.
     k = id(dataType)
     cls = _cached_cls.get(k)
-    if cls is None:
+    if cls is None or cls.__datatype is not dataType:
```

Comment:
Yes, there is a side effect that the id of dataType will be not recycled once we keep a reference of it in `cls`.

In the beginning, I thought the check is helping, actually, it's not. We can change it to an assert, or remove it.